### ISSUE TITLE
BUG: Speed issue with `setdiff1d`

### ISSUE THREAD

**Issue by philipGaudreau** [numpy/numpy#21804]
### Describe the issue:

I need to perform a set difference between 2 numpy arrays with `dtype=str`, so I'm using `np.setdiff1d`. However, it takes much more time than the `np.intersect1d` followed by `np.delete` function for some reason.

For ~340 000 texts in each array, it takes 48 minutes for `np.setdiff1d`, while it takes not even a second for `np.intersect1d` on my machine.

The dataset I'm using can be downloaded from _Kaggle_ at https://www.kaggle.com/datasets/rtatman/blog-authorship-corpus/download.

### Reproduce the code example:

```python
from time import time

import numpy as np
import pandas as pd

df = pd.read_csv("blogtext.csv")
arr1 = df[:df.shape[0]//2]['text'].to_numpy().flatten()
arr2 = df[df.shape[0]//2:]['text'].to_numpy().flatten()
arr1=np.unique(arr1)

# Method 1
start = time()
updated1 = np.setdiff1d(arr1, arr2)
end = time()
print(end-start)

# Method 2
start = time()
_, indices, _ = np.intersect1d(arr1, arr2, return_indices=True)
updated2 = np.delete(arr1, indices)
end = time()
print(end-start)

assert (updated1 == update2).all()
```

### NumPy/Python version information:

1.22.4 3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]
Running on Ubuntu 20.04.

Output of `lscpu`:
```
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
...
Thread(s) per core:              1
...
CPU MHz:                         1497.531
CPU max MHz:                     2650.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        5299.71
...
L1d cache:                       1.5 MiB
L1i cache:                       1.5 MiB
L2 cache:                        24 MiB
L3 cache:                        256 MiB
...
```

**Comment by anordin95** [comment]
Hi! I'm encountering this same issue. I wrote up a simple speed test suite. Here are the results for my machine.
For reference this is with `python3.11.4` and `numpy1.25.2`

Set difference runtimes using uuid's w/ size: 10000
fn: <function np_set_diff at 0x102aba0c0> ran for 9.1412s.
fn: <function python_list_diff at 0x1041f2020> ran for 5.5533s.
fn: <function python_set_diff at 0x1041f2160> ran for 0.0027s.
Set difference runtimes using int's w/ size: 10000
fn: <function np_set_diff at 0x102aba0c0> ran for 0.0015s.
fn: <function python_list_diff at 0x1041f2020> ran for 0.5716s.
fn: <function python_set_diff at 0x1041f2160> ran for 0.0013s.



```
import time
import numpy as np
import uuid
import random

def timer(fn):
  def time_fn(*args, **kwargs):
    fn_start_time = time.time()
    return_val = fn(*args, **kwargs)
    fn_end_time = time.time()
    print(f"fn: {fn} ran for {fn_end_time - fn_start_time:.4f}s.")

    return return_val
  return time_fn

@timer
def np_set_diff(arr_a, arr_b):
  return np.setdiff1d(arr_a, arr_b)

@timer
def python_list_diff(arr_a, arr_b):
  return [item for item in arr_a if item not in arr_b]

@timer
def python_set_diff(arr_a, arr_b):
  set_a = set(arr_a)
  set_b = set(arr_b)
  return list(set_a.difference(set_b))

if __name__ == "__main__":
  SIZE = int(1e4)
  
  print(f"Set difference runtimes using uuid's w/ size: {SIZE}")  
  arr_a = [uuid.uuid4() for _ in range(SIZE)]
  arr_b = [uuid.uuid4() for _ in range(SIZE)]
  np_set_diff(arr_a, arr_b)
  python_list_diff(arr_a, arr_b)
  python_set_diff(arr_a, arr_b)
  
  print(f"Set difference runtimes using int's w/ size: {SIZE}")  
  # Mimic, albeit for from perfectly, the unlikelihood of collisions observed in the uuid type.
  arr_a = [random.randint(0, SIZE*100) for _ in range(SIZE)]
  arr_b = [random.randint(0, SIZE*100) for _ in range(SIZE)]
  np_set_diff(arr_a, arr_b)
  python_list_diff(arr_a, arr_b)
  python_set_diff(arr_a, arr_b)
```

**Comment by charris** [comment]
Out of curiosity, how does it scale with size?

**Comment by anordin95** [comment]
Hey! In the section "some notable observed outputs", there are results for varying size from 1e4 to 1e7 and between int & uuid types. Does that section sort-of answer what you are curious to see?

**Comment by charris** [comment]
> Does that section sort-of answer what you are curious to see?

No.  I was curious about scaling with the problematic data type as it might show an inefficient algorithm. The other way to do it would be to read the code, but I was lazy :)

**Comment by anordin95** [comment]
Set difference runtimes using two arrays each with 100 uuid's
fn: <function np_set_diff at 0x113de8310> ran for 0.0023s.
fn: <function python_list_diff at 0x113dfc0d0> ran for 0.0009s.
fn: <function python_set_diff at 0x113dfc1f0> ran for 0.0000s.

Set difference runtimes using two arrays each with 1,000 uuid's
fn: <function np_set_diff at 0x113de8310> ran for 0.1360s.
fn: <function python_list_diff at 0x113dfc0d0> ran for 0.0881s.
fn: <function python_set_diff at 0x113dfc1f0> ran for 0.0003s.

Set difference runtimes using two arrays each with 10,000 uuid's
fn: <function np_set_diff at 0x113de8310> ran for 12.4210s.
fn: <function python_list_diff at 0x113dfc0d0> ran for 8.3075s.
fn: <function python_set_diff at 0x113dfc1f0> ran for 0.0029s.

Results look pretty relatively consistent for the uuid type regardless of scale. Can you explain more about what you were thinking? I'm not sure I follow  in this case how multiple data points for scale would provide clarity into an algorithm's efficiency.

Sidenote: times are generally a bit slower as I'm on a different, weaker machine.

**Comment by charris** [comment]
Not much data, but it looks like `python_set_diff` scales linearly with size, `np_set_diff` as the square.  That makes me suspect that `python_set_diff` is using a different algorithm.

**Comment by anordin95** [comment]
![rel_fig](https://github.com/numpy/numpy/assets/7705027/b3f9ca98-9f3f-4889-9701-e7692cb8d790)
![log_rel_fig](https://github.com/numpy/numpy/assets/7705027/55157771-856c-42c0-9cac-0c28fe9f0eff)

**Comment by charris** [comment]
@anordin95 Thanks. Looks pretty quadratic, which would be the naive way to implement the function. Looks like there is a fair amount of literature on the problem. I would guess using sorted data would be much faster on anything of substantial size. A log-log plot is the usual way to show the scaling.
